jetty-http2/http2-client/src/main/java/org/eclipse/jetty/http2/client/HTTP2ClientConnectionFactory.java

```diff
@@ -98,12 +101,52 @@ public void onOpen()
             Map<Integer, Integer> settings = listener.onPreface(getSession());
             if (settings == null)
                 settings = new HashMap<>();
-            settings.computeIfAbsent(SettingsFrame.INITIAL_WINDOW_SIZE, k -> client.getInitialStreamRecvWindow());
-            settings.computeIfAbsent(SettingsFrame.MAX_CONCURRENT_STREAMS, k -> client.getMaxConcurrentPushedStreams());
 
-            Integer maxFrameLength = settings.get(SettingsFrame.MAX_FRAME_SIZE);
-            if (maxFrameLength != null)
-                getParser().setMaxFrameLength(maxFrameLength);
+            // Below we want to populate any settings to send to the server
+            // that have a different default than what prescribed by the RFC.
+            // Changing the configuration is done when the SETTINGS is sent.
+
+            settings.compute(SettingsFrame.HEADER_TABLE_SIZE, (k, v) ->
+            {
+                if (v == null)
+                {
+                    v = client.getMaxDecoderTableSize();
+                    if (v == 4096)
+                        v = null;
+                }
+                return v;
+            });
+            settings.computeIfAbsent(SettingsFrame.MAX_CONCURRENT_STREAMS, k -> client.getMaxConcurrentPushedStreams());
+            settings.compute(SettingsFrame.INITIAL_WINDOW_SIZE, (k, v) ->
+            {
+                if (v == null)
+                {
+                    v = client.getInitialStreamRecvWindow();
+                    if (v == FlowControlStrategy.DEFAULT_WINDOW_SIZE)
+                        v = null;
+                }
+                return v;
+            });
+            settings.compute(SettingsFrame.MAX_FRAME_SIZE, (k, v) ->
+            {
+                if (v == null)
+                {
+                    v = client.getMaxFrameSize();
+                    if (v == Frame.DEFAULT_MAX_LENGTH)
+                        v = null;
+                }
+                return v;
+            });
+            settings.compute(SettingsFrame.MAX_HEADER_LIST_SIZE, (k, v) ->
+            {
+                if (v == null)
+                {
+                    v = client.getMaxResponseHeadersSize();
+                    if (v <= 0)
+                        v = null;
+                }
+                return v;
+            });
```

Comment:
I would have expected to see these values be set on the decoder after this like in the HTTP/3 equivalent.
Why is this not done here?

Reply:
Because in HTTP/2 the SETTINGS frame is sent only once, while here it can be sent multiple times and at any time.
Here it is only the place for the first, mandatory, SETTINGS frame, for which we need to provide sane initial values.

Because SETTINGS frame can be sent later by the application, the values on the decoder are set in `beforeSend()`.